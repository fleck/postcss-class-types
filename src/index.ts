import * as appRoot from "app-root-path";
import createSelectorParser from "postcss-selector-parser";
import nodePath from "path";
import fs from "fs-extra";
import postcssLib, { plugin } from "postcss";

export const defaultDirectory = nodePath.join(
  appRoot.toString(),
  "@types",
  "class-types"
);

export const classNamesDeclarationFilename = "classTypes.d.ts";

const parser = createSelectorParser(selectors => {
  selectors.each(selector => {
    if (selector.type !== "selector") {
      selector.remove();
      return;
    }

    selector.walk(node => {
      if (
        node.type !== "class" &&
        (node.type !== "combinator" || [">", "~"].includes(node.value))
      ) {
        node.remove();
      }
    });
  });
});

const autoGeneratedFileWarning =
  "// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.\n// prettier-ignore\n";

const initializer = ({ directory = "" } = {}) => {
  const dest = directory || defaultDirectory;

  const indexFileOutput = fs.outputFile(
    nodePath.join(dest, "index.d.ts"),
    `\
${autoGeneratedFileWarning}
declare module "class-types.macro" {
  // prettier-ignore
  export default function ct<T extends ClassTypes>(...a: T[]): T;
}
`
  );

  return async (root: postcssLib.Root) => {
    const classes = new Set<string>();

    classes.add("");

    root.walkRules(rule => {
      parser.processSync(rule.selector, {
        lossless: false,
        updateSelector: false,
      });

      parser
        .processSync(rule.selector, {
          lossless: false,
          updateSelector: false,
        })
        .trim()
        .replace(/(?<!\\)\./g, "")
        .replace(/\\/g, "")
        .replace(/,/g, "")
        .split(" ")
        .filter(Boolean)
        .forEach(cssClass => classes.add(cssClass));
    });

    return Promise.all([
      indexFileOutput,
      fs.outputFile(
        nodePath.join(dest, classNamesDeclarationFilename),
        `${autoGeneratedFileWarning}type ClassTypes = "${Array.from(
          classes
        ).join('" | "')}"\n`
      ),
    ]);
  };
};

export type options = Parameters<typeof initializer>[0];

export const postcss = true;

const postcssClassTypes = plugin<options>("postcss-class-types", initializer);

export default postcssClassTypes;

module.exports = postcssClassTypes;

module.exports.defaultDirectory = defaultDirectory;

module.exports.classNamesDeclarationFilename = classNamesDeclarationFilename;
