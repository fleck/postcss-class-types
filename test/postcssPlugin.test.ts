import postcss from "postcss";
import { default as plugin, options } from "../src/";
import fs from "fs-extra";
import path from "path";

async function run(input: string, opts: options) {
  let result = await postcss([plugin(opts)]).process(input, {
    from: undefined,
  });

  expect(result.warnings()).toHaveLength(0);
}

const typesDirectory = path.join(__dirname, "..", "@types");

const cleanupFiles = () => {
  fs.removeSync(typesDirectory);
};

beforeEach(cleanupFiles);
afterEach(cleanupFiles);

const classNamesPath = path.join(typesDirectory, "ct.macro", "classNames.d.ts");

test("generates types for classes", async () => {
  await run(".mt-1{ } .mt-2{}", {});

  expect((await fs.readFile(classNamesPath)).toString()).toMatchSnapshot();
});

test("does not include ids, pseudo selectors, or attributes", async () => {
  await run(
    '#app .mt-1{ } #app .mt-2:hover{} [data-reach-slider-input][data-orientation="horizontal"] {}',
    {}
  );

  expect((await fs.readFile(classNamesPath)).toString()).toMatchSnapshot();
});

test("supports custom directories", async () => {
  const directory = path.join(typesDirectory, "custom");

  await run("#app .mt-1{ } #app .mt-2:hover{}", { directory });

  expect(
    (await fs.readFile(path.join(directory, "classNames.d.ts"))).toString()
  ).toMatchSnapshot();
});

test("groups", async () => {
  const directory = path.join(typesDirectory, "custom");

  await run("#td .group:hover .group-hover\\:text-gray-900{}", { directory });

  expect(
    (await fs.readFile(path.join(directory, "classNames.d.ts"))).toString()
  ).toMatchSnapshot();
});

test("space", async () => {
  const directory = path.join(typesDirectory, "custom");

  await run(
    `.space-y-2>:not(template)~:not(template) {
       --space-y-reverse:0!important;margin-top: calc(0.5rem*(1 - var(--space-y-reverse)))!important;
       margin-bottom: calc(0.5rem*var(--space-y-reverse))!important
     }`,
    { directory }
  );

  expect(
    (await fs.readFile(path.join(directory, "classNames.d.ts"))).toString()
  ).toMatchSnapshot();
});
