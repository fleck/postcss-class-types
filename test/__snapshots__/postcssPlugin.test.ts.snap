// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comma 1`] = `
"// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
type ClassTypes = \\"\\""
`;

exports[`does not include ids, pseudo selectors, or attributes 1`] = `
"// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
type ClassTypes = \\"mt-1\\" | \\"mt-2\\""
`;

exports[`generates types for classes 1`] = `
"// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
type ClassTypes = \\"mt-1\\" | \\"mt-2\\""
`;

exports[`groups 1`] = `
"// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
type ClassTypes = \\"group\\" | \\"group-hover:text-gray-900\\""
`;

exports[`space 1`] = `
"// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
type ClassTypes = \\"space-y-2\\""
`;

exports[`supports custom directories 1`] = `
"// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
type ClassTypes = \\"mt-1\\" | \\"mt-2\\""
`;
